WalletWasabi/Stores/IndexStore.cs

```diff
@@ -23,153 +20,81 @@ namespace WalletWasabi.Stores;
+	/// <remarks>Guarded by <see cref="IndexLock"/>.</remarks>
+	private Task InitializeFiltersNoLockAsync(CancellationToken cancellationToken)
 	{
 		try
 		{
-			if (MatureIndexFileManager.Exists())
-			{
-				using (BenchmarkLogger.Measure(LogLevel.Debug, "MatureIndexFileManager loading"))
-				{
-					int i = 0;
-					using StreamReader sr = MatureIndexFileManager.OpenText();
-
-					if (!sr.EndOfStream)
-					{
-						while (true)
-						{
-							i++;
-							cancellationToken.ThrowIfCancellationRequested();
-							string? line = await sr.ReadLineAsync(CancellationToken.None).ConfigureAwait(false);
-
-							if (line is null)
-							{
-								break;
-							}
-
-							ProcessLineNoLock(line, enqueue: false);
-						}
-					}
-
-					Logger.LogDebug($"Loaded {i} lines from the mature index file.");
-				}
-			}
-		}
-		catch (Exception ex) when (ex is not OperationCanceledException)
-		{
-			// We found a corrupted entry. Stop here.
-			// Delete the corrupted file.
-			// Do not try to autocorrect, because the internal data structures are throwing events that may confuse the consumers of those events.
-			Logger.LogError("Mature index got corrupted. Deleting both mature and immature index...");
-			MatureIndexFileManager.DeleteMe();
-			ImmatureIndexFileManager.DeleteMe();
-			throw;
-		}
+			// TODO: Replace the message with a new one. Staying with this one to allow comparisons.
+			using IDisposable _ = BenchmarkLogger.Measure(LogLevel.Debug, "MatureIndexFileManager loading");
 
-		cancellationToken.ThrowIfCancellationRequested();
+			int i = 0;
 
-		try
-		{
-			if (ImmatureIndexFileManager.Exists())
+			// Read last N filters. There is no need to read all of them.
+			foreach (FilterModel filter in IndexStorage.FetchLast(n: 5000))
 			{
-				foreach (var line in await ImmatureIndexFileManager.ReadAllLinesAsync(cancellationToken).ConfigureAwait(false)) // We can load ImmatureIndexFileManager to the memory, no problem.
+				i++;
+
+				if (!TryProcessFilterNoLock(filter, enqueue: false))
 				{
-					ProcessLineNoLock(line, enqueue: true);
-					cancellationToken.ThrowIfCancellationRequested();
+					throw new InvalidOperationException("Index file inconsistency detected.");
 				}
+
+				cancellationToken.ThrowIfCancellationRequested();
 			}
+
+			Logger.LogDebug($"Loaded {i} lines from the mature index file.");
 		}
 		catch (Exception ex) when (ex is not OperationCanceledException)
 		{
-			// We found a corrupted entry. Stop here.
-			// Delete the corrupted file.
-			// Do not try to autocorrect, because the internal data structures are throwing events that may confuse the consumers of those events.
-			Logger.LogError("Immature index got corrupted. Deleting it...");
-			ImmatureIndexFileManager.DeleteMe();
+			// We found a corrupted entry. Clear the corrupted database and stop here.
+			Logger.LogError("Filter index got corrupted. Clearing the filter index...");
+			IndexStorage.Clear();
```

Comment:
Why should only catch those exceptions that we can handle, otherwise it makes no sense to do anything. Imagine we get a `SQLiteException "table not found"` then  `IndexStorage.Clear();` is pointless. In this case the only exception that seems to be candidate to be handled is `InvalidOperationException`. 

Note: The concept of immature filters describes the number of blocks that we can resist a blockchain reorg. it is an agreed number that we respects everywhere in the code.

I am not against removing them all but only if we find an addressable problem and not something like a NRE

Reply:
> Ah, I see. I modified behavior here by deleting complete table instead of last X records. Right?

No, you didn't modify the behavior because you are deleting more than the minimum. I think it is okay.

> Though I don't think we should ever hit this catch clause (say I replace Exception for InvalidOperationException).

I agree. If you replace the `catch(Exception)` by `catch(InvalidOperationException)` then we should hit it in extremely rare situations.